freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtV1Service.java

```diff
@@ -143,13 +149,30 @@ private String serviceAdd(String serviceName, String hostname, String realm, Fre
         RPCResponse<com.sequenceiq.freeipa.client.model.Service> response;
         com.sequenceiq.freeipa.client.model.Service service;
         try {
-            response = ipaClient.addService(serviceName + "/" + hostname + "@" + realm);
-            if (response == null || response.getResult() == null) {
-                return null;
+            String canonicalPrincipal = serviceName + "/" + hostname + "@" + realm;
+            try {
+                response = ipaClient.addService(canonicalPrincipal);
+                if (response == null || response.getResult() == null) {
+                    return null;
+                }
+                service = (com.sequenceiq.freeipa.client.model.Service) response.getResult();
+            } catch (FreeIpaClientException e) {
+                if (!Optional.ofNullable(e.getCause())
+                        .filter(JsonRpcClientException.class::isInstance)
+                        .map(JsonRpcClientException.class::cast)
+                        .map(JsonRpcClientException::getCode)
+                        .filter(c -> c == DUPLICATE_ENTRY_ERROR_CODE)
+                        .isPresent()) {
+                    throw e;
+                }
+
+                response = ipaClient.showService(canonicalPrincipal);
+                if (response == null || response.getResult() == null) {
+                    return null;
```

Comment:
wouldn't make it sense that instead of this many `return null` we would let propagate up a `FreeIpaClientException` and the caller would try-catch all FPCE and convert to the appropriate `KeytabCreationException`. Or if we expect this to be missing an Optional wrapper would make it more readable that it can return without value.

Reply:
I cleaned this up. Your recommendation about not needing to have the whole `RPCResponse` helped make it cleaner too.